storage/pin/pin_test.go

```diff
@@ -0,0 +1,509 @@
+func TestWalker(t *testing.T) {
+	sizes := []int{1, 4095, 4096, 4097, 123456}
+	for i := range sizes {
+		p, f, closeFunc := getPinApiAndFileStore(t)
+		defer closeFunc()
+
+		data := testutil.RandomBytes(1, sizes[i])
+		hash := uploadFile(t, f, data, false)
+
+		addrs, err := f.GetAllReferences(context.TODO(), bytes.NewReader(data))
+		if err != nil {
+			t.Fatalf("Error getting original chunks of the file")
+		}
+
+		// Function to collect the walked chunks
+		walkedChunks := make(map[string]uint64)
+		var lock = sync.RWMutex{}
+		walkerFunction := func(ref storage.Reference) error {
+			chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+			lock.Lock()
+			defer lock.Unlock()
+			walkedChunks[hex.EncodeToString(chunkAddr)] = 0
+			return nil
+		}
+		p.walkChunksFromRootHash(hex.EncodeToString(hash), true, "", walkerFunction)
+
+		// Check if the number of chunks and chunk addresses match
+		if len(addrs) != len(walkedChunks) {
+			t.Fatalf("Expected number of chunks to be %d, but is %d", len(addrs), len(walkedChunks))
+		}
+		for _, hash := range addrs {
+			if _, ok := walkedChunks[hex.EncodeToString(hash)]; !ok {
+				t.Fatalf("Expected chunk %s not present", hash)
+			}
+		}
+	}
+}
+
+// TestListPinInfo tests the ListPinFiles command by pinning and unpinning a collection
+// twice and check if this gets reflected properly in the data structure
+func TestListPinInfo(t *testing.T) {
+	p, f, closeFunc := getPinApiAndFileStore(t)
+	defer closeFunc()
+
+	hash := uploadCollection(t, p, f, false)
+
+	// Pin the hash for the first time
+	err := p.PinFiles(hex.EncodeToString(hash), false, "")
+	if err != nil {
+		t.Fatalf("Could not pin " + err.Error())
+	}
+
+	// Get the list of pinned files by calling the ListPinFiles command
+	pinInfo := p.ListPinFiles()
+
+	// Check if the uploaded collection is in the list files data structure
+	fileInfo, ok := pinInfo[hex.EncodeToString(hash)]
+	if !ok {
+		t.Fatalf("uploaded collection not pinned")
+	}
+	if fileInfo.PinCounter != 1 {
+		t.Fatalf("pincounter expected is 1 got is %d", fileInfo.PinCounter)
+	}
+	if fileInfo.IsRaw {
+		t.Fatalf("IsRaw expected is false got is true")
+	}
+
+	// Pin it once more and check if the counters increases
+	err = p.PinFiles(hex.EncodeToString(hash), false, "")
+	fileInfo, ok = pinInfo[hex.EncodeToString(hash)]
+	if err != nil {
+		t.Fatalf("Could not pin " + err.Error())
+	}
+	if fileInfo.PinCounter != 2 {
+		t.Fatalf("pincounter expected is 2 got is %d", fileInfo.PinCounter)
+	}
+
+	// Unpin it and check if the counter decrements
+	err = p.UnpinFiles(hex.EncodeToString(hash), "")
+	if err != nil {
+		t.Fatalf("Could not unpin " + err.Error())
+	}
+	fileInfo, ok = pinInfo[hex.EncodeToString(hash)]
+	if !ok {
+		t.Fatalf("collection totally unpinned")
+	}
+	if fileInfo.PinCounter != 1 {
+		t.Fatalf("pincounter expected is 1 got is %d", fileInfo.PinCounter)
+	}
+
+	// Unpin it final time and the entry should not be there
+	err = p.UnpinFiles(hex.EncodeToString(hash), "")
```

Comment:
hash is already an address or byte slice. the API should not handle `string`s but concrete types

Reply:
How do you expect a user to input address in bytes?  There is no "rpc command" yet client which converts the user input to []bytes.

Reply:
the low level API should be handling the concrete types, but then the API type you expose through the `APIs()` can take a `string` type through an exposed wrapper method, validate the input, then call the underlying API method. Not gonna block on this - as said you can PR this later